partiql-eval/src/main/kotlin/org/partiql/eval/internal/Compiler.kt

```diff
@@ -23,6 +23,7 @@ import org.partiql.eval.internal.operator.rex.ExprCase
 import org.partiql.eval.internal.operator.rex.ExprCast
```

Comment:
Wouldn't you need to add scopes to every Rel and Expr?

Reply:
Short answer -- no. This PR is an attempt to use both the Volcano Model as well as the Variables Environment (stack). 

In a traditional program (non-SQL), you might need to place values on the stack to invoke a function, but we're modeling data flow through the transfer of records -- volcano model. So, most expressions don't need access to this "global" environment of scopes. And, they don't need to push/pop scopes. Same goes for relations.

We can kind of think of new scopes as new concatenations of the current variables environment -- and we don't need to concatenate the variables environment for every Rel/Expr. The PartiQL Spec talks about it briefly:
> In particular, the WHERE’s condition is evaluated once for each input binding tuple b in Bin
WHERE. In general, each evaluation is done within the bindings environment (ρ0, ρ‖b), i.e., the concatenation of the binding tuple ρ (where ρ is the binding environment of the SFW query) with the binding tuple b that has the variables of the FROM clause

It kind of describes the situation where `ρ` is the scope of variables defined outside of the SFW query, and `b` is the binding tuple being passed from the FILTER's input (SCAN). Assuming this is a SELECT .. FROM .. WHERE, I'm going to make the assumption that the evaluation of the projection is done within the bindings environment `(ρ0, ρ‖b1)` (and not `(ρ0, ρ‖b‖b1)`) where `b1` is the binding tuple produced from the filter. This intuitively makes sense because of how you can't reference a table's columns in a projection list when there is an aggregation.

So, with the concept of both `Local` and `Outer`, it helps differentiate when you are referencing something from the input record (volcano) vs when things are coming from outer variable environments (stack).

So, for example:
```
SELECT
  (SELECT t1.a + t2.a FROM t2)
FROM t1
```

Will be transformed into:
```
Plan = Select(
  Project(                                                      // Call this Project_1
    projection = Coerce(SubPlan_1)
    input = Scan(t1)
  )
)

SubPlan_1 = Select(
  Project(                                                      // Call this Project_2
    projection = Outer(0,0).a + Local(0).a
    input = Scan(t2)
  )
)
```

We know that, at the moment we are evaluating the projection in Project_1, our variables environment is realistically `<set of variables defined outside of query_1> || <output of scan(t1)>`. But, upon evaluating Project_2's projection, the variables environment would look like: `<set of variables defined outside of query_2> || <output of scan(t2)>` == `<set of variables defined outside of query_1> || <output of scan(t1)> || <output of scan(t2)>`.